pkg/distributor/distributor_test.go

```diff
@@ -2037,59 +2036,140 @@ func TestDistributor_Push_ShouldGuaranteeShardingTokenConsistencyOverTheTime(t *
+//						skipLabelNameValidation: tc.skipLabelNameValidationCfg,
+//					})
+//					req := mockWriteRequest([]labels.Labels{tc.inputLabels}, 42, 100000, histogram)
+//					req.SkipLabelNameValidation = tc.skipLabelNameValidationReq
+//					_, err := ds[0].Push(ctx, req)
+//					if tc.errExpected {
+//						fromError, _ := status.FromError(err)
+//						assert.Equal(t, tc.errMessage, fromError.Message())
+//					} else {
+//						assert.Nil(t, err)
+//					}
+//				})
+//			}
+//		}
+//	}
 func TestDistributor_Push_LabelNameValidation(t *testing.T) {
-	t.Parallel()
-	inputLabels := labels.Labels{
-		{Name: model.MetricNameLabel, Value: "foo"},
-		{Name: "999.illegal", Value: "baz"},
-	}
 	ctx := user.InjectOrgID(context.Background(), "user")
 
 	tests := map[string]struct {
 		inputLabels                labels.Labels
 		skipLabelNameValidationCfg bool
 		skipLabelNameValidationReq bool
+		useUTF8Validation          bool
 		errExpected                bool
 		errMessage                 string
 	}{
 		"label name validation is on by default": {
-			inputLabels: inputLabels,
+			inputLabels: labels.Labels{
+				{Name: model.MetricNameLabel, Value: "foo"},
+				{Name: "999.illegal", Value: "baz"},
+			},
 			errExpected: true,
 			errMessage:  `sample invalid label: "999.illegal" metric "foo{999.illegal=\"baz\"}"`,
 		},
 		"label name validation can be skipped via config": {
-			inputLabels:                inputLabels,
+			inputLabels: labels.Labels{
+				{Name: model.MetricNameLabel, Value: "foo"},
+				{Name: "999.illegal", Value: "baz"},
+			},
 			skipLabelNameValidationCfg: true,
 			errExpected:                false,
 		},
 		"label name validation can be skipped via WriteRequest parameter": {
-			inputLabels:                inputLabels,
+			inputLabels: labels.Labels{
+				{Name: model.MetricNameLabel, Value: "foo"},
+				{Name: "999.illegal", Value: "baz"},
+			},
 			skipLabelNameValidationReq: true,
 			errExpected:                false,
 		},
+		"UTF-8 validation allows Unicode label names": {
+			inputLabels: labels.Labels{
+				{Name: model.MetricNameLabel, Value: "foo"},
+				{Name: "Cortex_😃", Value: "baz"},
+			},
+			useUTF8Validation: true,
+			errExpected:       false,
+		},
 	}
 
 	for testName, tc := range tests {
-		tc := tc
 		for _, histogram := range []bool{true, false} {
-			histogram := histogram
-			t.Run(fmt.Sprintf("%s, histogram=%s", testName, strconv.FormatBool(histogram)), func(t *testing.T) {
-				t.Parallel()
+			t.Run(fmt.Sprintf("%s, histogram=%v", testName, histogram), func(t *testing.T) {
+				if tc.useUTF8Validation {
+					// nolint:staticcheck // SA1019: using deprecated NameValidationScheme intentionally for legacy validation testing
+					model.NameValidationScheme = model.UTF8Validation
+				} else {
+					// nolint:staticcheck // SA1019: using deprecated NameValidationScheme intentionally for legacy validation testing
+					model.NameValidationScheme = model.LegacyValidation
```

Comment:
We shouldn't be modifying `model.NameValidationScheme` directly in the test case. Let's use the new config we added instead.